### PR TITLE
Adding /quarantine and /unquarantine

### PR DESCRIPTION
**Pull request by Zabuzard**
### Overview

This adds the `/quarantine` and `/unquarantine`  commands. They act mostly as preparation for #394 , but should also aid as help for mods and staff to manually resolve situations.

Quarantining is similar to muting, but will have subtle differences in how it is handled in the server, permission-wise. For example, a quarantined user might only be given access to a special `#quarantine_zone` channel, where he can talk to the moderators. Also, they will have less permissions than muted users, for example not being able to view any other channel anymore.

Quarantines are permanent, but the code supports temporary quarantines as well (this will be interesting for the scam blocker).

apply

quarantine_zone

unquarantine

### Details

This is mostly a copy&paste&edit of **mute**. Touched stuff:

Commands:
* `QuarantineCommand`
* `UnquarantineCommand`
* added enum values in `ModerationAction`
* added role helpers in `ModerationUtils`

To auto-revoke temporary quarantines:
* `TemporaryQuarantineAction`
* edits in `TemporaryModerationRoutine`

To reapply the role after rejoin:
* merged existing `RejoinMuteListener` with quarantine into more general `RejoinModerationRoleListener`

### Config

This changed the config. See the template:
```json
"quarantinedRolePattern": "Quarantined"
```

### CODE REVIEW

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/RejoinModerationRoleListener.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+            // User was never e.g. muted
+            return false;
+        }
+
+        Optional<ActionRecord> lastRevokeAction = actionsStore.findLastActionAgainstTargetByType(
+                member.getGuild().getIdLong(), member.getIdLong(), moderationRole.revokeAction);
+        if (lastRevokeAction.isEmpty()) {
+            // User was never e.g. unmuted
+            return isActionEffective(lastRevokeAction.orElseThrow());
+        }
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/RejoinModerationRoleListener.java:86]: I'm extremely confused.

if there's no action

throw?

I mean, lastRevokeAction is empty? So orElseThrow throws?

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/QuarantineCommand.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        String dmMessage =
+                """
+                        Hey there, sorry to tell you but unfortunately you have been put under quarantine in the server %s.
+                        This means you can no longer interact with anyone in the server until you have been unquarantined again.
+                        If you think this was a mistake, or the reason no longer applies, please contact a moderator or admin of the server.
+                        The reason for the quarantine is: %s
+                        """
+                    .formatted(guild.getName(), reason);
+        return event.getJDA()
+            .openPrivateChannelById(target.getId())
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/QuarantineCommand.java:77]:
```suggestion
            .openPrivateChannelById(target.getIdLong())
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnquarantineCommand.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    private static RestAction<Boolean> sendDm(@NotNull ISnowflake target, @NotNull String reason,
+            @NotNull Guild guild, @NotNull GenericEvent event) {
+        String dmMessage = """
+                Hey there, you have been put out of quarantine in the server %s.
+                This means you can now interact with others in the server again.
+                The reason for the unquarantine is: %s
+                """.formatted(guild.getName(), reason);
+        return event.getJDA()
+            .openPrivateChannelById(target.getId())
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnquarantineCommand.java:76]:
```suggestion
            .openPrivateChannelById(target.getIdLong())
```

**Comment by Zabuzard** [comment]
> SonarCloud Quality Gate failed.    Quality Gate failed
> 
> [Bug](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=CODE_SMELL) [3 Code Smells](https://sonarcloud.io/project/issues?id=Together-Java_TJ-Bot&pullRequest=398&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=Together-Java_TJ-Bot&pullRequest=398&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=Together-Java_TJ-Bot&pullRequest=398&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=Together-Java_TJ-Bot&pullRequest=398&metric=new_duplicated_lines_density&view=list) [11.1% Duplication](https://sonarcloud.io/component_measures?id=Together-Java_TJ-Bot&pullRequest=398&metric=new_duplicated_lines_density&view=list)

Smells are code duplication. The duplication is between mute and quarantine. While it is correct, I think it might be better to have the duplication. Overly coupling both actions would result in a pretty messy code that is hard to view isolated.

After all, the two commands serve different purposes.

An argument could be made to abstract both to some sort of "apply role" commands. Maybe it is worth pursuing this direction, but the claimed duplication is actually mostly on the user-responses and command layout, which wouldnt be covered by such a merge anyways...

So for the same reason we accepted the duplication between ban, warn and mute, I am voting to accept it here as well.

**Review by Tais993 — CHANGES_REQUESTED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/RejoinModerationRoleListener.java (hunk 1 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/RejoinModerationRoleListener.java:86]: Actually, I might have done a typo during refactoring here. (This class already existed and was renamed + some extras)

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/RejoinModerationRoleListener.java]: Awesome spot. fixed.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/QuarantineCommand.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            @NotNull Guild guild, @NotNull GenericEvent event) {
+        String dmMessage =
+                """
+                        Hey there, sorry to tell you but unfortunately you have been put under quarantine in the server %s.
+                        This means you can no longer interact with anyone in the server until you have been unquarantined again.
+                        If you think this was a mistake, or the reason no longer applies, please contact a moderator or admin of the server.
+                        The reason for the quarantine is: %s
+                        """
+                    .formatted(guild.getName(), reason);
+        return event.getJDA()
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/QuarantineCommand.java:76]:
```suggestion
                    .formatted(guild.getName(), reason);
                    
        return event.getJDA()
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/QuarantineCommand.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    @SuppressWarnings({"BooleanMethodNameMustStartWithQuestion", "MethodWithTooManyParameters"})
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @Nullable Member target, @NotNull CharSequence reason, @NotNull Guild guild,
+            @NotNull Interaction event) {
+        if (!ModerationUtils.handleRoleChangeChecks(
+                ModerationUtils.getQuarantinedRole(guild, config).orElse(null), ACTION_VERB, target,
+                bot, author, guild, hasRequiredRole, reason, event)) {
+            return false;
+        }
+        if (Objects.requireNonNull(target)
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/QuarantineCommand.java:127]:
```suggestion
        }
        
        if (Objects.requireNonNull(target)
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnquarantineCommand.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    private static RestAction<Boolean> sendDm(@NotNull ISnowflake target, @NotNull String reason,
+            @NotNull Guild guild, @NotNull GenericEvent event) {
+        String dmMessage = """
+                Hey there, you have been put out of quarantine in the server %s.
+                This means you can now interact with others in the server again.
+                The reason for the unquarantine is: %s
+                """.formatted(guild.getName(), reason);
+        return event.getJDA()
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnquarantineCommand.java:75]:
```suggestion
                """.formatted(guild.getName(), reason);
                
        return event.getJDA()
```

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnquarantineCommand.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    @SuppressWarnings({"BooleanMethodNameMustStartWithQuestion", "MethodWithTooManyParameters"})
+    private boolean handleChecks(@NotNull Member bot, @NotNull Member author,
+            @Nullable Member target, @NotNull CharSequence reason, @NotNull Guild guild,
+            @NotNull Interaction event) {
+        if (!ModerationUtils.handleRoleChangeChecks(
+                ModerationUtils.getQuarantinedRole(guild, config).orElse(null), ACTION_VERB, target,
+                bot, author, guild, hasRequiredRole, reason, event)) {
+            return false;
+        }
+        if (Objects.requireNonNull(target)
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnquarantineCommand.java:126]:
```suggestion
        }
        
        if (Objects.requireNonNull(target)
```

**Review by Tais993 — COMMENTED** [review]
Mr Styling has returned, besides that;

I'm fine with ignoring the code duplication, so 👍 from me

**Review by Tais993 — DISMISSED** [review]

**Review by Tais993 — APPROVED** [review]

**Comment by Zabuzard** [comment]
One week has passed, merging

**Pull request merged**